File: SRC.NET/Dialog.vb
Option Strict Off
Option Explicit On
Friend Class Dialog
	
	' Copyright (C) 1997-2012 Kei Sakamoto / Inui Tetsuyuki
	'Invalid_string_refer_to_original_code
	'Invalid_string_refer_to_original_code
	'Invalid_string_refer_to_original_code
	
	'繝繧､繧｢繝ｭ繧ｰ縺ｮ繧ｯ繝ｩ繧ｹ
	
	'Invalid_string_refer_to_original_code
	Private intMessageNum As Object
	'Invalid_string_refer_to_original_code
	Private strName() As String
	'Invalid_string_refer_to_original_code
	Private strMessage() As String
	
	'Invalid_string_refer_to_original_code
	'UPGRADE_NOTE: Class_Initialize は Class_Initialize_Renamed にアップグレードされました。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="A9E4979A-37FA-4718-9994-97DD76ED70A7"' をクリックしてください。
	Private Sub Class_Initialize_Renamed()
		'UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
		intMessageNum = 0
		ReDim strName(0)
		ReDim strMessage(0)
	End Sub
	Public Sub New()
		MyBase.New()
		Class_Initialize_Renamed()
	End Sub
	
	'繧ｯ繝ｩ繧ｹ縺ｮ霑ｽ蜉
	'UPGRADE_NOTE: Class_Terminate は Class_Terminate_Renamed にアップグレードされました。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="A9E4979A-37FA-4718-9994-97DD76ED70A7"' をクリックしてください。
	Private Sub Class_Terminate_Renamed()
		ReDim strName(0)
		ReDim strMessage(0)
	End Sub
	Protected Overrides Sub Finalize()
		Class_Terminate_Renamed()
		MyBase.Finalize()
	End Sub
	
	'Invalid_string_refer_to_original_code
	Public Sub AddMessage(ByRef Name As String, ByRef Message As String)
		'UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
		intMessageNum = intMessageNum + 1
		ReDim Preserve strName(intMessageNum)
		ReDim Preserve strMessage(intMessageNum)
		'UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
		strName(intMessageNum) = Name
		'UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
		strMessage(intMessageNum) = Message
	End Sub
	
	'Invalid_string_refer_to_original_code
	Public ReadOnly Property Count() As Short
		Get
			'UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
			Count = intMessageNum
		End Get
	End Property
	
	'Invalid_string_refer_to_original_code
	Public ReadOnly Property IsAvailable(ByVal u As Unit, ByVal ignore_condition As Boolean) As Boolean
		Get
			Dim pname, pname2 As String
			Dim i, j As Short
			Dim mpnickname, mpname, mtype As String
			
			With u.MainPilot
				mpname = .Name
				mpnickname = .Nickname
				mtype = .MessageType
			End With
			
			For i = 1 To Count
				pname = strName(i)
				
				'Invalid_string_refer_to_original_code
				If Left(pname, 1) = "@" Then
					pname = Mid(pname, 2)
					For j = 1 To UBound(SelectedPartners)
						With SelectedPartners(j)
							If .CountPilot > 0 Then
								'Invalid_string_refer_to_original_code
								With .MainPilot
									If pname <> .Name And InStr(pname, .Name & "(") <> 1 And pname <> .Nickname And InStr(pname, .Nickname & "(") <> 1 Then
										GoTo NextPartner
									End If
								End With
								
								'Invalid_string_refer_to_original_code
								If Not ignore_condition Then
									'Invalid_string_refer_to_original_code_
									'Invalid_string_refer_to_original_code_
									'Invalid_string_refer_to_original_code_
									'Or .IsConditionSatisfied("豺ｷ荵ｱ") _
									'Then
									'UPGRADE_ISSUE: 前の行を解析できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="82EBB1AE-1FCB-4FEF-9E6C-8736A316F8A7"' をクリックしてください。
									IsAvailable = False
									Exit Property
								End If
							End If
							
							'Invalid_string_refer_to_original_code
							GoTo NextMessage
						End With
					Next 
				End If
				'End With
NextPartner: 
			Next 
			
			IsAvailable = False
			Exit Property
			'End If
			
			'Invalid_string_refer_to_original_code
			If InStr(pname, "(") > 0 Then
				If Not PDList.IsDefined2(pname) And NPDList.IsDefined2(pname) Then
					'Invalid_string_refer_to_original_code
					For j = 2 To Len(pname)
						If Mid(pname, Len(pname) - j, 1) = "(" Then
							pname2 = Left(pname, Len(pname) - j - 1)
							Exit For
						End If
					Next 
					
					'Invalid_string_refer_to_original_code
					If PDList.IsDefined2(pname2) Or NPDList.IsDefined2(pname2) Then
						'Invalid_string_refer_to_original_code
						pname = pname2
					End If
				End If
			End If
			
			'Invalid_string_refer_to_original_code
			If pname = mpname Then
				GoTo NextMessage
			End If
			If pname = mpnickname Then
				GoTo NextMessage
			End If
			If pname = mtype Then
				GoTo NextMessage
			End If
			
			'Invalid_string_refer_to_original_code
			If NPDList.IsDefined(pname) Then
				If IsGlobalVariableDefined("IsAway(" & pname & ")") Then
					IsAvailable = False
					Exit Property
				End If
				GoTo NextMessage
			End If
			
			If PDList.IsDefined(pname) Then
				'Invalid_string_refer_to_original_code
				
				'Invalid_string_refer_to_original_code
				If Not PList.IsDefined(pname) Then
					IsAvailable = False
					Exit Property
				End If
				
				With PList.Item(pname)
					'Invalid_string_refer_to_original_code
					If .Away Then
						IsAvailable = False
						Exit Property
					End If
					
					'Invalid_string_refer_to_original_code
					If Not ignore_condition And Not .Unit Is Nothing Then
						With .Unit
							'Invalid_string_refer_to_original_code_
							'Invalid_string_refer_to_original_code_
							'Invalid_string_refer_to_original_code_
							'Or .IsConditionSatisfied("豺ｷ荵ｱ") _
							'Then
							'UPGRADE_ISSUE: 前の行を解析できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="82EBB1AE-1FCB-4FEF-9E6C-8736A316F8A7"' をクリックしてください。
							IsAvailable = False
							Exit Property
						End With
					End If
				End With
			End If
			'End With
			'End If
			
NextMessage: 
			'Next
			
			IsAvailable = True
		End Get
	End Property
	
	'Invalid_string_refer_to_original_code
	Public Function Name(ByVal idx As Short) As String
		Name = strName(idx)
	End Function
	
	'Invalid_string_refer_to_original_code
	Public Function Message(ByVal idx As Short) As String
		Message = strMessage(idx)
	End Function
End Class

File: SRC.NET/Dialog.cs
﻿using System;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace Project1
{
    internal class Dialog
    {

        // Copyright (C) 1997-2012 Kei Sakamoto / Inui Tetsuyuki
        // 本プログラムはフリーソフトであり、無保証です。
        // 本プログラムはGNU General Public License(Ver.3またはそれ以降)が定める条件の下で
        // 再頒布または改変することができます。

        // ダイアログのクラス

        // ダイアログに含まれるメッセージ数
        private object intMessageNum;
        // メッセージの話者
        private string[] strName;
        // メッセージ
        private string[] strMessage;

        // クラスの初期化
        // UPGRADE_NOTE: Class_Initialize は Class_Initialize_Renamed にアップグレードされました。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="A9E4979A-37FA-4718-9994-97DD76ED70A7"' をクリックしてください。
        private void Class_Initialize_Renamed()
        {
            // UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
            intMessageNum = 0;
            strName = new string[1];
            strMessage = new string[1];
        }

        public Dialog() : base()
        {
            Class_Initialize_Renamed();
        }

        // クラスの追加
        // UPGRADE_NOTE: Class_Terminate は Class_Terminate_Renamed にアップグレードされました。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="A9E4979A-37FA-4718-9994-97DD76ED70A7"' をクリックしてください。
        private void Class_Terminate_Renamed()
        {
            strName = new string[1];
            strMessage = new string[1];
        }

        ~Dialog()
        {
            Class_Terminate_Renamed();
        }

        // ダイアログにメッセージを追加
        public void AddMessage(ref string Name, ref string Message)
        {
            // UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
            intMessageNum = Operators.AddObject(intMessageNum, 1);
            Array.Resize(ref strName, Conversions.ToInteger(intMessageNum + 1));
            Array.Resize(ref strMessage, Conversions.ToInteger(intMessageNum + 1));
            // UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
            strName[Conversions.ToInteger(intMessageNum)] = Name;
            // UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
            strMessage[Conversions.ToInteger(intMessageNum)] = Message;
        }

        // ダイアログに含まれるメッセージ数
        public short Count
        {
            get
            {
                short CountRet = default;
                // UPGRADE_WARNING: オブジェクト intMessageNum の既定プロパティを解決できませんでした。 詳細については、'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"' をクリックしてください。
                CountRet = Conversions.ToShort(intMessageNum);
                return CountRet;
            }
        }

        // ダイアログに使われているパイロットが全て利用可能か判定
        public bool get_IsAvailable(Unit u, bool ignore_condition)
        {
            bool IsAvailableRet = default;
            string pname, pname2 = default;
            short i, j;
            string mpnickname, mpname, mtype;
            {
                var withBlock = u.MainPilot();
                mpname = withBlock.Name;
                mpnickname = withBlock.get_Nickname(false);
                mtype = withBlock.MessageType;
            }

            var loopTo = Count;
            for (i = 1; i <= loopTo; i++)
            {
                pname = strName[i];

                // 合体技のパートナーが指定されている場合
                if (Strings.Left(pname, 1) == "@")
                {
                    pname = Strings.Mid(pname, 2);
                    var loopTo1 = (short)Information.UBound(Commands.SelectedPartners);
                    for (j = 1; j <= loopTo1; j++)
                    {
                        {
                            var withBlock1 = Commands.SelectedPartners[j];
                            if (withBlock1.CountPilot() > 0)
                            {
                                // パートナーの名前と一致する？
                                {
                                    var withBlock2 = withBlock1.MainPilot();
                                    if ((pname ?? "") != (withBlock2.Name ?? "") & Strings.InStr(pname, withBlock2.Name + "(") != 1 & (pname ?? "") != (withBlock2.get_Nickname(false) ?? "") & Strings.InStr(pname, withBlock2.get_Nickname(false) + "(") != 1)
                                    {
                                        goto NextPartner;
                                    }
                                }

                                // 喋れるかどうかチェック
                                if (!ignore_condition)
                                {
                                    object argIndex1 = "睡眠";
                                    object argIndex2 = "麻痺";
                                    object argIndex3 = "石化";
                                    object argIndex4 = "恐怖";
                                    object argIndex5 = "沈黙";
                                    object argIndex6 = "混乱";
                                    if (withBlock1.IsConditionSatisfied(ref argIndex1) | withBlock1.IsConditionSatisfied(ref argIndex2) | withBlock1.IsConditionSatisfied(ref argIndex3) | withBlock1.IsConditionSatisfied(ref argIndex4) | withBlock1.IsConditionSatisfied(ref argIndex5) | withBlock1.IsConditionSatisfied(ref argIndex6))
                                    {
                                        IsAvailableRet = false;
                                        return default;
                                    }
                                }

                                // メッセージは使用可能
                                goto NextMessage;
                            }
                        }

                        NextPartner:
                        ;
                    }

                    IsAvailableRet = false;
                    return default;
                }

                // 表情パターンが指定されている？
                if (Strings.InStr(pname, "(") > 0)
                {
                    bool localIsDefined21() { object argIndex1 = pname; var ret = SRC.PDList.IsDefined2(ref argIndex1); return ret; }

                    bool localIsDefined22() { object argIndex1 = pname; var ret = SRC.NPDList.IsDefined2(ref argIndex1); return ret; }

                    if (!localIsDefined21() & localIsDefined22())
                    {
                        // 括弧部分を削除
                        var loopTo2 = (short)Strings.Len(pname);
                        for (j = 2; j <= loopTo2; j++)
                        {
                            if (Strings.Mid(pname, Strings.Len(pname) - j, 1) == "(")
                            {
                                pname2 = Strings.Left(pname, Strings.Len(pname) - j - 1);
                                break;
                            }
                        }

                        // 表情パターンかどうか判定
                        bool localIsDefined2() { object argIndex1 = pname2; var ret = SRC.NPDList.IsDefined2(ref argIndex1); return ret; }

                        object argIndex7 = pname2;
                        if (SRC.PDList.IsDefined2(ref argIndex7) | localIsDefined2())
                        {
                            // 表情パターンとみなす
                            pname = pname2;
                        }
                    }
                }

                // メインパイロットは常に存在
                if ((pname ?? "") == (mpname ?? ""))
                {
                    goto NextMessage;
                }

                if ((pname ?? "") == (mpnickname ?? ""))
                {
                    goto NextMessage;
                }

                if ((pname ?? "") == (mtype ?? ""))
                {
                    goto NextMessage;
                }

                // ノンパイロットはLeaveしていない限り常に存在
                object argIndex8 = pname;
                if (SRC.NPDList.IsDefined(ref argIndex8))
                {
                    string argvname = "IsAway(" + pname + ")";
                    if (Expression.IsGlobalVariableDefined(ref argvname))
                    {
                        IsAvailableRet = false;
                        return default;
                    }

                    goto NextMessage;
                }

                object argIndex10 = pname;
                if (SRC.PDList.IsDefined(ref argIndex10))
                {
                    // パイロットの場合

                    // パイロットが作成されていない？
                    bool localIsDefined() { object argIndex1 = pname; var ret = SRC.PList.IsDefined(ref argIndex1); return ret; }

                    if (!localIsDefined())
                    {
                        IsAvailableRet = false;
                        return default;
                    }

                    object argIndex9 = pname;
                    {
                        var withBlock3 = SRC.PList.Item(ref argIndex9);
                        // Leave中？
                        if (withBlock3.Away)
                        {
                            IsAvailableRet = false;
                            return default;
                        }

                        // 喋れるかどうかチェック
                        if (!ignore_condition & withBlock3.Unit is object)
                        {
                            {
                                var withBlock4 = withBlock3.Unit;
                                if (withBlock4.IsConditionSatisfied("睡眠") | withBlock4.IsConditionSatisfied("麻痺") | withBlock4.IsConditionSatisfied("石化") | withBlock4.IsConditionSatisfied("恐怖") | withBlock4.IsConditionSatisfied("沈黙") | withBlock4.IsConditionSatisfied("混乱"))
                                {
                                    IsAvailableRet = false;
                                    return default;
                                }
                            }
                        }
                    }
                }

                NextMessage:
                ;
            }

            IsAvailableRet = true;
            return IsAvailableRet;
        }

        // idx番目のメッセージの話者
        public string Name(short idx)
        {
            string NameRet = default;
            NameRet = strName[idx];
            return NameRet;
        }

        // idx番目のメッセージ
        public string Message(short idx)
        {
            string MessageRet = default;
            MessageRet = strMessage[idx];
            return MessageRet;
        }
    }
}